File: new/python/19_remove_nth_node_from_end_of_list_two_pointers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # idea: two pointers
        # create a dummy node and make left point to it as the previous value of head
        # make the right point to head
        # move right pointer exactly n steps forward as the first step
        # then move both left and right pointer until right reaches the end.
        # now left will be exactly one before the node to be removed, manipulate next pointers to remove the intended node
        dummy = ListNode(0, head)
        left = dummy  # this will save us from bugs, if we want to remove the head
        right = head

        while n != 0:
            right = right.next
            n -= 1

        while right:
            left = left.next
            right = right.next

        left.next = left.next.next

        return dummy.next  # returning the head pointer
        # time: O(n)
        # space: O(1)

File: new/cpp/19_remove_nth_node_from_end_of_list_two_pointers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // idea: two pointers (dummy node)
        // use a dummy node that points to head (a node before head)
        // use left and right pointer. 
        // make left point to dummy
        // and right point to head. 
        // now move right by exactly n nodes to the right
        // now move both left and right until right reaches nullptr
        // if right is at the end, then left is one before the node to be removed,
        // use left->next and left->next->next to remove the node and return 
        // !!!! dummy->next;

        ListNode* dummy = new ListNode(0, head);
        ListNode* left = dummy;
        ListNode* right = head;

        while (n > 0) {
            right = right->next;
            n--;
        }

        while (right != nullptr) {
            right = right->next;
            left = left->next;
        }

        left->next = left->next->next;
        return dummy->next;
    }
    // time: O(n)
    // space: O(1)
};
